x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenizer.java

```diff
@@ -121,6 +121,17 @@ public BasicTokenizer(boolean isLowerCase, boolean isTokenizeCjkChars) {
         return processedTokens;
     }
 
+    private int findIndexBeforePunctuation(String token) {
+        int i = token.length() - 1;
+        while (i > 0) {
```

Comment:
This method can never return 0.  Is that a sign of a bug, or is it impossible that it could ever get called for a string like `a!!!!!`?  Either change `>` to `>=` or add a comment explaining why not.

Also, if passed a string that doesn't contain any punctuation, like `aaaa`, it will return the index of the last character.  But that's not before punctuation.  If that's intentional then it's probably best rename the method to `findLastNonPunctuationIndex`.

Reply:
There is no bug because of the fact that a token where only the first character is non-punctuation, won't get split any further. But I think the `finldIndexBeforePunctuation` function doesn't say what it promises. I have refactored the code so the function is no `findLastNonPunctuationIndex` and I'm also only calling it now if the last character is punctuation to avoid wasted work.